types/react-redux/index.d.ts

```diff
@@ -15,12 +15,12 @@ type Store<S> = Redux.Store<S>;
 type Dispatch<S> = Redux.Dispatch<S>;
 type ActionCreator<A> = Redux.ActionCreator<A>;
 
-export interface DispatchProp<T> {
-  dispatch: Dispatch<T>
+export interface DispatchProp<S> {
+  dispatch: Dispatch<S>;
 }
 
-interface ComponentDecorator<TMergedProps> {
-    <TOwnProps>(component: Component<(TOwnProps & TMergedProps) | TOwnProps>): ComponentClass<TOwnProps>;
+interface ComponentDecorator<TMergedProps, TOwnProps> {
+    <T extends TOwnProps>(component: Component<T & TMergedProps>): ComponentClass<T>;
```

Comment:
Hey @blakeembrey - I'm trying to understand some type issues we're having (related to https://github.com/DefinitelyTyped/DefinitelyTyped/issues/16990#issuecomment-308256211). We don't fully understand why the `<T extends TOwnProps>` is needed. It seems like below should be correct in terms of what react-redux connect returns:
```
interface ComponentDecorator<TMergedProps, TOwnProps> {
  (component: Component<TOwnProps & TMergedProps>): ComponentClass<TOwnProps>;
}
```
When we change it to that the problem in that linked comment seems to go away, because it can correctly infer that the connected component only accepts ownProps, not all the props of the dumb / presentation component.
I've added a test in my fork which passes if I change to that:
https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...ritwik:react-redux/fix-ownprop-inference
However, the problem is that now the tests for untyped connect calls don't work: 
```
react-redux-tests.tsx(286,1): error TS2322: Type 'ComponentClass<{}>' is not assignable to type 'ComponentClass<TestProp>'.
  Type '{}' is not assignable to type 'TestProp'.
    Property 'property1' is missing in type '{}'.
react-redux-tests.tsx(289,23): error TS2339: Property 'property1' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<{}, ComponentState>> & Readonly<{ childr...'.
react-redux-tests.tsx(301,39): error TS2345: Argument of type '(props: HelloMessageProps) => Element' is not assignable to parameter of type 'Component<DispatchProp<any>>'.
  Type '(props: HelloMessageProps) => Element' is not assignable to type 'StatelessComponent<DispatchProp<any>>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'DispatchProp<any> & { children?: ReactNode; }' is not assignable to type 'HelloMessageProps'.
        Property 'name' is missing in type 'DispatchProp<any> & { children?: ReactNode; }'.
react-redux-tests.tsx(303,40): error TS2339: Property 'name' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<{}, ComponentState>> & Readonly<{ childr...'.
```
So my assumption is that it's needed for the cases where you're less explicit about the ownprop types?